### PR TITLE
👪 Family model and resources

### LINKED ISSUE #184: Create Family Model
TODO

### LINKED ISSUE #207: Family endpoint
Write a /family endpoint that supports basic CRUD operations:

- [ ] GET /families - return families from dataservice
- [ ]  GET /families/{kf_id} - Get a family by id
- [ ]  POST /families - Creates a family in the dataservice
- [ ]  PUT /families/{kf_id} - Updates a family by id
- [ ]  DELETE /families/{kf_id} - Deletes a family

### PR DESCRIPTION
**Pull request by dankolbman**
Adds family model and resources, replacing the current `family_id` field in participant.

Closes #184, Closes #207

### CODE REVIEW

**Review by parimalak — COMMENTED** [review]
Can you update erd and migrations?

File: dataservice/api/family/models.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from dataservice.api.common.model import Base
+from dataservice.api.participant.models import Participant
+
+
+class Family(db.Model, Base):
+    """
+    Family entity.
+
+    :param kf_id: Unique id given by the Kid's First DCC
+    :param external_id: Name given to participant by contributor
```
Comment (znatty22) [inline dataservice/api/family/models.py]: Change to something like `Name or id given to family by contributor`

**Comment by dankolbman** [comment]
Updated with new erd and migrations.

File: dataservice/api/participant/schemas.py [hunk 2]
```diff
@@ -6,20 +6,35 @@
...
+        exclude = ('study', 'family')
 
     _links = ma.Hyperlinks({
         'self': ma.URLFor(Meta.resource_url, kf_id='<kf_id>'),
-        'collection': ma.URLFor(Meta.collection_url)
+        'collection': ma.URLFor(Meta.collection_url),
+        'study': ma.URLFor('api.studies', kf_id='<study_id>'),
     })
+
+    def dump(self, p, *args, **kwargs):
```
Comment (znatty22) [inline dataservice/api/participant/schemas.py:27]: Sigh its unfortunate that this is still an issue in flask-ma [.](https://github.com/marshmallow-code/flask-marshmallow/issues/68) I tried looking at the source code to see if we can handle this in a more generic way and thought of a potential fix. However, I don't think its worth it to implement bc i'm unsure of how the fix could affect query params when we decide to implement that in our API.

File: dataservice/api/participant/schemas.py [hunk 3]
```diff
@@ -6,20 +6,35 @@
...
+        a link if there is before dumping, then remove link after the dump
+        """
+        if kwargs['many'] is False and p.family_id is not None:
+            self.fields['_links'].schema['family'] = (
+                ma.URLFor('api.families', kf_id='<family_id>')
+            )
+            data = super(ParticipantSchema, self).dump(p, *args, **kwargs)
+            del self.fields['_links'].schema['family']
+            return data
+        data = super(ParticipantSchema, self).dump(p, *args, **kwargs)
```
Comment (znatty22) [inline dataservice/api/participant/schemas.py]: Minor thing.. you could probably just return `super(ParticipantSchema, self).dump(p, *args, **kwargs)` directly

**Review by znatty22 — COMMENTED** [review]
Looks good just a couple minor comments

**Comment by dankolbman** [comment]
Also snuck in a fix to exclude the uuid from some of the resources that had hand-defined exclude fields in the schema.

**Review by znatty22 — APPROVED** [review]
🥇

**Review by parimalak — APPROVED** [review]
👍

**Pull request merged**